File: ios/alier_dev/alier/http/data/Envelope.swift
/*
Copyright 2024 Suredesigns Corp.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import Foundation

enum BinaryEncoding {
    case Base64
    case QuotedPrintable
}

protocol HTTPBody {
    func toByteArray() -> ByteArray
    func equals(other: Any?) -> Bool
}

class NoContent: HTTPBody {
    func equals(other: Any?) -> Bool {
        if (other == nil) { return false }
        return self === other as! NoContent || other is NoContent
    }

    func toByteArray() -> ByteArray {
        return ByteArray("".utf8)
    }
}

class Text: HTTPBody {
    private var body: String
    init(body: String) {
        self.body = body
    }
    func toString() -> String {
        return self.body
    }

    func toByteArray() -> ByteArray {
        return ByteArray(self.body.utf8)
    }
    
    func equals(other: Any?) -> Bool {
        if other is Text {
            return self.body == (other as! Text).body
        }
        return false
    }
}

class Binary: HTTPBody {
    private var body: ByteArray
    init(body: ByteArray) {
        self.body = body
    }
    
    static func fromString(s: String) -> Binary {
        if (s.hasPrefix("data:")) {
            let range = s.range(of: ";base64,")
            if range == nil {
                let startIndex = s.index(s.startIndex, offsetBy: 13)
                let data = s[startIndex..<s.endIndex].data(using: .utf8)
                return Binary(body: (data?.encodedHexadecimals())!)
            } else {
                let index = s.index(s.startIndex, offsetBy: 5)
                return Binary(body: ByteArray(s[index].utf8))
            }
        } else {
            return Binary(body: ByteArray(s.utf8))
        }
    }

    func toString() -> String {
        return toString(encoding: BinaryEncoding.Base64)
    }
    func toString(encoding: BinaryEncoding) -> String {
        switch (encoding) {
        case BinaryEncoding.Base64:
            let data = Data(body)
            let encoded = Data(base64Encoded: data)
            return String(data: encoded!, encoding: .utf8)!
        case BinaryEncoding.QuotedPrintable:
            return QuotedPrintable.encodeToString(input: body)
        }
    }

    func toByteArray() -> ByteArray {
        return self.body
    }
    func equals(other: Any?) -> Bool {
        if self === other as! Binary { return true }
        if other is Binary {
            return self.body == (other as! Binary).body
        } else {
            return false
        }
    }
}

class NestedEnvelopes: HTTPBody {
    private var body: Array<Envelope>
    private var contentType: String
    private var boundary: String
    
    init(body: Array<Envelope>, contentType: String, boundary: String) {
        self.body = body
        self.contentType = contentType
        self.boundary = boundary
    }
    
    func toByteArray() -> ByteArray {
        let delimiter = "--\(self.boundary)".data(using: .ascii)?.encodedHexadecimals()
        let crlf = "\r\n".data(using: .ascii)?.encodedHexadecimals()
        let chunks: NaiveByteBuffer = NaiveByteBuffer(initialCapacity: delimiter!.count * 2)

        for envelope in self.body {
            chunks
                .add(chunk: delimiter!)
                .add(chunk: crlf!)
                .add(chunk: envelope.toByteArray())
        }
        chunks
            .add(chunk: delimiter!)
            .add(chunk: ("--".data(using: .ascii)?.encodedHexadecimals())!)
            .add(chunk: crlf!)

        return chunks.toByteArray()
    }
    
    func equals(other: Any?) -> Bool {
        if self === other as! NestedEnvelopes { return true }
        if other is NestedEnvelopes {
            let otherBody = (other as! NestedEnvelopes).body
            var res = true;
            for i in 0 ..< self.body.count {
                res = self.body[i].equals(other: otherBody[i])
                if !res { break; }
            }
            return res
        } else {
            return false
        }
    }
}

class NestedEnvelope: HTTPBody {
    private var body: Envelope
    private var boundary: String
    
    init(body: Envelope, boundary: String) {
        self.body = body
        self.boundary = boundary
    }
    
    func toByteArray() -> ByteArray {
        let delimiter = "--\(self.boundary)".data(using: .ascii)?.encodedHexadecimals()
        let crlf = "\r\n".data(using: .ascii)?.encodedHexadecimals()
        let chunks: NaiveByteBuffer = NaiveByteBuffer(initialCapacity: delimiter!.count * 2)

        chunks
            .add(chunk: delimiter!)
            .add(chunk: crlf!)
            .add(chunk: self.body.toByteArray())
            .add(chunk: delimiter!)
            .add(chunk: ("--".data(using: .ascii)?.encodedHexadecimals())!)
            .add(chunk: crlf!)

        return chunks.toByteArray()
    }
    
    func equals(other: Any?) -> Bool {
        if other is Envelope {
            return self.body.equals(other: (other as! NestedEnvelope).body)
        }
        return false
    }
}


class Envelope {
    var headers: Headers
    var body: HTTPBody
    
    init(headers: Headers, body: HTTPBody) {
        self.headers = headers
        self.body = body
    }
    
    func toByteArray() -> ByteArray {
        let crlf = "\r\n".data(using: .ascii)?.encodedHexadecimals()
        let chunks: NaiveByteBuffer = NaiveByteBuffer()

        chunks
            .add(chunk: self.headers.toByteArray())
            .add(chunk: crlf!)
            .add(chunk: self.body.toByteArray())

        return chunks.toByteArray()
    }
    
    func equals(other: Any?) -> Bool {
        if self === other as! Envelope { return true }
        if other is Envelope {
            if !self.headers.equals(other: (other as! Envelope).headers) { return false }
            return self.body.equals(other: (other as! Envelope).body)
        } else {
            return false
        }
    }
}

File: android/alier_dev/alier/src/main/java/com/suredesigns/alier/net/http/data/Envelope.kt
/*
Copyright 2024 Suredesigns Corp.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package com.suredesigns.alier.net.http.data

import android.util.Base64
import java.nio.charset.Charset


enum class BinaryEncoding {
    Base64,
    QuotedPrintable,
}

sealed class Body {
    abstract fun toByteArray(): ByteArray
}

class NoContent: Body() {
    override fun equals(other: Any?): Boolean {
        return this === other || other is NoContent
    }

    override fun hashCode(): Int {
        return javaClass.hashCode()
    }

    override fun toByteArray(): ByteArray {
        return ByteArray(0)
    }
}

data class Text(val body: String): Body() {
    override fun toString(): String {
        return body
    }

    override fun toByteArray(): ByteArray {
        return body.toByteArray(Charsets.UTF_8)
    }
    fun toByteArray(charset: Charset): ByteArray {
        return body.toByteArray(charset)
    }
}

data class Binary(val body: ByteArray): Body() {
    companion object {
        fun fromString(s: String): Binary {
            return Binary(if (s.startsWith("data:")) {
                val encoding_pos = s.indexOf(";base64,", 5)
                if (encoding_pos < 0) {
                    Base64.decode(s.substring(encoding_pos + ";base64,".length), Base64.NO_WRAP or Base64.URL_SAFE)
                } else {
                    s.substring(5).toByteArray()
                }
            } else {
                s.toByteArray()
            })
        }
    }
    constructor(body: String): this(fromString(body).body)

    override fun toString(): String {
        return toString(BinaryEncoding.Base64)
    }
    fun toString(encoding: BinaryEncoding): String {
        return when (encoding) {
            BinaryEncoding.Base64 -> Base64.encodeToString(body, Base64.NO_WRAP or Base64.URL_SAFE)
            BinaryEncoding.QuotedPrintable -> QuotedPrintable.encodeToString(body)
        }
    }

    override fun toByteArray(): ByteArray {
        return body
    }
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Binary) return false

        return body.contentEquals(other.body)
    }

    override fun hashCode(): Int {
        return body.contentHashCode()
    }
}

data class NestedEnvelopes(val body: Array<Envelope>, val contentType: String, val boundary: String): Body() {
    override fun toByteArray(): ByteArray {
        val delimiter = "--${boundary}".toByteArray(Charsets.US_ASCII)
        val crlf = "\r\n".toByteArray(Charsets.US_ASCII)
        val chunks = NaiveByteBuffer(delimiter.size * 2)

        for (envelope in body) {
            chunks
                .add(delimiter)
                .add(crlf)
                .add(envelope.toByteArray())
        }
        chunks
            .add(delimiter)
            .add("--".toByteArray(Charsets.US_ASCII))
            .add(crlf)

        return chunks.toByteArray()
    }
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is NestedEnvelopes) return false

        return body.contentEquals(other.body)
    }

    override fun hashCode(): Int {
        return body.contentHashCode()
    }
}
data class NestedEnvelope(val body: Envelope, val boundary: String): Body() {
    override fun toByteArray(): ByteArray {
        val delimiter = "--${boundary}".toByteArray(Charsets.US_ASCII)
        val crlf = "\r\n".toByteArray(Charsets.US_ASCII)
        val chunks = NaiveByteBuffer(delimiter.size * 2)

        chunks
            .add(delimiter)
            .add(crlf)
            .add(body.toByteArray())
            .add(delimiter)
            .add("--".toByteArray(Charsets.US_ASCII))
            .add(crlf)

        return chunks.toByteArray()
    }
}


data class Envelope(val headers: Headers, val body: Body) {
    fun toByteArray(): ByteArray {
        val crlf = "\r\n".toByteArray(Charsets.US_ASCII)
        val chunks = NaiveByteBuffer()

        chunks
            .add(headers.toByteArray())
            .add(crlf)
            .add(body.toByteArray())

        return chunks.toByteArray()
    }
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Envelope) return false
        if (headers != other.headers) return false
        if (body != other.body) return false

        return true
    }

    override fun hashCode(): Int {
        var result = headers.hashCode()
        result = 31 * result + body.hashCode()
        return result
    }

}
